### PR TITLE
Improve perf of float and double parsing in TextBuffer

### PR DESCRIPTION
**Pull request by pjfanning**
* just an outline - not for merging
* relates to #1229 
* In TextBuffer, when parsing floats and doubles, avoid creating Strings first
* copies the logic of contentsAsString into the float and double methods so that we can best handle the various scenarios
  * if _resultString already exists, use it
  * if _resultArray already exists, parse from the array instead of creating a String
  * in the case that multiple segments are needed, we might as well fall back to contentsAsString in this case because we need to build a String with a StringBuilder anyway 

@cowtowncoder is this a long the right lines? If so, I can add tests. If we get this done for floats and doubles, we can come back to make similar changes for other number types. 

I'm aware that this does not address the deferred number parsing in ParserBase (where numbers can be stored as `_numberString` until we know what type we need). Storing the number as the wrong number type and later converting it to a different number type can lead to number conversion issues and can be more expensive.

### CODE REVIEW

File: src/main/java/com/fasterxml/jackson/core/io/NumberInput.java [hunk 1]
```diff
@@ -400,6 +400,32 @@ public static double parseDouble(final String s, final boolean useFastParser) th
         return useFastParser ? JavaDoubleParser.parseDouble(s) : Double.parseDouble(s);
     }
 
+    /**
+     * @param array a char array containing a number to parse
+     * @param useFastParser whether to use {@code FastDoubleParser}
+     * @return closest matching double
+     * @throws NumberFormatException if value cannot be represented by a double
+     * @since v2.18
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/io/NumberInput.java:408]: I think we use plain `2.18` for version

**Comment by cowtowncoder** [comment]
@pjfanning Yes, this is what I had in mind. I'll have a look and suggest some changes. Since we are not (yet) changing calling from `ParserBase`, I think this could actually go in 2.17, even.

I also think that we can remove some of deferral because calls to, say, `getFloatValue()` does explicitly establish target type -- but this is easy enough to verify with more time wrt 2.18 once 2.17 is out.
But that question is separate from work to make `TextBuffer` pass more optimal representation.

File: src/main/java/com/fasterxml/jackson/core/util/TextBuffer.java [hunk 2]
```diff
@@ -536,6 +536,37 @@ public char[] contentsAsArray() throws IOException {
      */
     public double contentsAsDouble(final boolean useFastParser) throws NumberFormatException {
         try {
+            if (_resultString == null) {
+                // Has array been requested? Can make a shortcut, if so:
+                if (_resultArray != null) {
+                    validateFPLength(_resultArray.length);
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/util/TextBuffer.java:542]: Not needed: `StreamReadConstraints.validateFPLength()` is called from `ParserBase.resetFloat()` as part of decoding. `ReaderBasedJsonParser, line 1465 f.ex calls:

```
        _textBuffer.resetWithShared(_inputBuffer, startPtr, len);
        // And there we have it!
        return resetFloat(neg, intLen, fractLen, expLen);
```

so this is already taken care of

File: src/main/java/com/fasterxml/jackson/core/util/TextBuffer.java [hunk 3]
```diff
@@ -536,6 +536,37 @@ public char[] contentsAsArray() throws IOException {
...
+                    return NumberInput.parseDouble(_resultArray, useFastParser);
+                } else {
+                    // Do we use shared array?
+                    if (_inputStart >= 0) {
+                        if (_inputLen < 1) {
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/util/TextBuffer.java:547]: Similarly, decoder has already confirmed this won't be the case, no need to verify.

File: src/main/java/com/fasterxml/jackson/core/util/ReadConstrainedTextBuffer.java [hunk 4]
```diff
@@ -26,4 +26,13 @@ protected void validateStringLength(int length) throws StreamConstraintsExceptio
     {
         _streamReadConstraints.validateStringLength(length);
     }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void validateFPLength(int length) throws StreamConstraintsException
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/util/ReadConstrainedTextBuffer.java:34]: Won't be needed, as per my comments (parsers will call `validateFPLength()` around same time as setting `TextBuffer` contents; and this only if number decoded is lexically valid (will not reset to empty String).

File: src/main/java/com/fasterxml/jackson/core/util/TextBuffer.java (hunk 3 above)
Comment (pjfanning) [reply src/main/java/com/fasterxml/jackson/core/util/TextBuffer.java:547]: when I removed this I got a NPE in a new test (empty TextBuffer)

Comment (cowtowncoder) [reply src/main/java/com/fasterxml/jackson/core/util/TextBuffer.java:546]: Right but we are not building a general purpose functionality: Parsers will never call (and are specifically not to ever call) method without populating buffer with lexically valid number.

File: src/test/java/com/fasterxml/jackson/core/util/TestTextBuffer.java [hunk 5]
```diff
@@ -192,4 +196,62 @@ public void testGetSizeFinishCurrentSegmentAndResetWith() throws Exception {
...
+    }
+
+    public void testContentsAsFloatFastParser() throws IOException {
+        TextBuffer textBuffer = new TextBuffer(null);
+        textBuffer.resetWithString("1.2345678");
+        assertEquals(1.2345678f,  textBuffer.contentsAsFloat(true));
+    }
+
+    public void testEmptyContentsAsDouble() {
+        TextBuffer textBuffer = new TextBuffer(null);
```
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/core/util/TestTextBuffer.java:226]: Not legal test as per usage patterns: not just for empty case but more generally, parsers will tokenize numbers using JSON (etc) pattern that describes valid numbers.
This simplifies decoding in cases decoder knows it need not validate input (such as our optimized `int`/`long` decoding logic).

Comment (pjfanning) [reply src/test/java/com/fasterxml/jackson/core/util/TestTextBuffer.java:243]: If `new TextBuffer(null)` is illegal, can we update the constructor to throw an exception?

Comment (cowtowncoder) [reply src/test/java/com/fasterxml/jackson/core/util/TestTextBuffer.java:243]: Yes, that'd make sense

Comment (cowtowncoder) [reply src/test/java/com/fasterxml/jackson/core/util/TestTextBuffer.java:243]: Actually no n/m, that's `BufferRecycler` -- `null` is accepted.

**Review by cowtowncoder — APPROVED** [review]

**Comment by cowtowncoder** [comment]
I think I'll go ahead and merge this: it will be needed to tackle #1229.

File: src/main/java/com/fasterxml/jackson/core/util/TextBuffer.java [hunk 6]
```diff
@@ -525,15 +524,41 @@ public char[] contentsAsArray() throws IOException {
...
+        // that do not require allocation first
+
+        // except _resultString first since it works best with JDK (non-fast parser)
+        if (_resultString != null) {
+            return NumberInput.parseDouble(_resultString, useFastParser);
+        }
+        if (_inputStart >= 0) { // shared?
+            return NumberInput.parseDouble(_inputBuffer, _inputStart, _inputLen, useFastParser);
+        }
+        if (_currentSize == 0) { // all content in current segment!
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/util/TextBuffer.java:554]: This was a bug, actually -- should use `_hasSegments` instead! Fixed in #1313.

File: src/main/java/com/fasterxml/jackson/core/util/TextBuffer.java [hunk 7]
```diff
@@ -574,14 +599,41 @@ public float contentsAsFloat() throws NumberFormatException {
...
+        // that do not require allocation first
+
+        // except _resultString first since it works best with JDK (non-fast parser)
+        if (_resultString != null) {
+            return NumberInput.parseFloat(_resultString, useFastParser);
+        }
+        if (_inputStart >= 0) { // shared?
+            return NumberInput.parseFloat(_inputBuffer, _inputStart, _inputLen, useFastParser);
+        }
+        if (_currentSize == 0) { // all content in current segment!
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/core/util/TextBuffer.java:629]: Same here, fixed in #1313

**Pull request merged**